The project's primary objective is to create a Restaurant Recommender Application designed to assist users in locating dining establishments that align with their tastes. Through the implementation of machine learning techniques, I have gathered genuine restaurant reviews and established a sophisticated recommender system.

In my project, I conducted an in-depth analysis of restaurant reviews from Vancouver. Throughout the project, I gathered data from web sources, constructed a comprehensive data frame, established a robust recommender system, and developed a user-friendly application to utilize the recommendation system seamlessly.

__Please note: this is notebook 2 of 3.__
    
In this notebook I combined reviews, rating and restaurant info data in one dataframe. I conducted data cleaning and exploratory data analysis (EDA) to prepare the data for subsequent analysis and modeling.

In [4]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import re

In [2]:
import session_info
session_info.show()

### Review

I'll begin by downloading the CSV file containing the scraped reviews and then proceed with the data cleaning process.

In [11]:
df_review = pd.read_csv('/Users/evgenijkucukov/Desktop/Brainstation/Portfolio/Restaurant_reviews_project/df_review_1')
df_review

restaurant  \
0                   Freshslice Pizza   
1                   Freshslice Pizza   
2                   Freshslice Pizza   
3                   Freshslice Pizza   
4                   Freshslice Pizza   
...                              ...   
18635  1050. Beach Ave Bar and Grill   
18636  1050. Beach Ave Bar and Grill   
18637  1050. Beach Ave Bar and Grill   
18638  1050. Beach Ave Bar and Grill   
18639  1050. Beach Ave Bar and Grill   

                                                  review  
0      Simar arora is very nice and sweet girl who se...  
1      I went to the Granville and 2400 block Vancouv...  
2      We were in a hurry to grab a snack. Pizza was ...  
3      Freshslice Pizza is hands down one of the best...  
4      I met there with their staff girl named Kiran,...  
...                                                  ...  
18635  Stopped in for lunch while I was walking the S...  
18636  Great food and beer in a perfect spot looking ...  
18637  We dine here often. And I almost always get a ...  
18638  The Beach Ave bar and grill is lovely spot to ...  
18639  My wife & I, who live in northern BC, rented a...  

[18640 rows x 2 columns]

In [12]:
# Check how many values are missing
df_review.isna().sum()

restaurant    0
review        0
dtype: int64

There are not missing values.

It's possible to observe that some restaurant names lack associated rating numbers. Additionally, we've identified duplicated restaurants resulting from web page advertisements. To address this, we should proceed to remove these particular restaurants from the dataset.

In [13]:
# delete ads reataurants (wo rate numbers)
df_review = df_review[df_review['restaurant'].str.contains(r'\d')]

After completing the data cleaning process, my next step is to merge all the dataframes based on the restaurant name as the common identifier. To ensure a smooth merging process, I plan to omit the rating numbers, as they may introduce inconsistencies during the merge.

In [14]:
df_review['restaurant'] = df_review['restaurant'].str.replace(r'^\d+\. ', '',regex=True)
df_review

/var/folders/0z/fdtwx1zs7lg98s4zb6q_8_4h0000gn/T/ipykernel_4151/2616152244.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_review['restaurant'] = df_review['restaurant'].str.replace(r'^\d+\. ', '',regex=True)


restaurant  \
26     Hydra Estiatorio Mediterranean   
27     Hydra Estiatorio Mediterranean   
28     Hydra Estiatorio Mediterranean   
29     Hydra Estiatorio Mediterranean   
30     Hydra Estiatorio Mediterranean   
...                               ...   
18635         Beach Ave Bar and Grill   
18636         Beach Ave Bar and Grill   
18637         Beach Ave Bar and Grill   
18638         Beach Ave Bar and Grill   
18639         Beach Ave Bar and Grill   

                                                  review  
26     Wonderful fresh exotic Greek food that is deli...  
27     Food was delicious, packed full of flavors. Th...  
28     Hello, We would like to say thank you for the ...  
29     Happy hour was amazing. Buck a shuck oysters a...  
30     Roman was the best waiter! he checked up on us...  
...                                                  ...  
18635  Stopped in for lunch while I was walking the S...  
18636  Great food and beer in a perfect spot looking ...  
18637  We dine here often. And I almost always get a ...  
18638  The Beach Ave bar and grill is lovely spot to ...  
18639  My wife & I, who live in northern BC, rented a...  

[16715 rows x 2 columns]

In [18]:
# let's check duplicates
duplicates = df_review[df_review.duplicated()]
duplicates

restaurant                                             review
128      Seaside Provisions  Thank you so much for the wonderful review! It...
185      Atlas Steak + Fish                      Thank you for the kind words!
307    Freebird Table & Bar  Thank you kindly for taking the time to write ...
317    Freebird Table & Bar  Thank you for the wonderful review! It was a p...
932     Seasons in the Park  Thank you for taking the time to provide us wi...
...                     ...                                                ...
18430      Freshslice Pizza  Thanks for taking the time to leave us a revie...
18440      Freshslice Pizza  We're thrilled to hear that you had a great ex...
18442      Freshslice Pizza  We're thrilled to hear that you had a great ex...
18444      Freshslice Pizza  We're thrilled to hear that you had a great ex...
18611          Boston Pizza  Thanks for the 5-star review! We're glad to he...

[104 rows x 2 columns]

In [19]:
# let's remove duplicated raws
df_review = df_review.drop_duplicates()

In [21]:
# sanity check
duplicates_1 = df_review[df_review.duplicated()]
duplicates_1

Empty DataFrame
Columns: [restaurant, review]
Index: []

There ara not duplicated values.

In [22]:
# let's check unique values
df_review.nunique()

restaurant     1009
review        16606
dtype: int64

In the cleaned dataframe, we have a total of 1,009 restaurants and 16,606 reviews.

Now let's repeat the same steps with rating and restaurant info data.

### Rating

In [24]:
df_rating = pd.read_csv('/Users/evgenijkucukov/Desktop/Brainstation/Portfolio/Restaurant_reviews_project/df_rating_1')
df_rating

restaurant  rating
0            Kozak Ukrainian Restaurant     4.5
1     1. Hydra Estiatorio Mediterranean     5.0
2                       2. Black + Blue     4.5
3                    3. Alouette Bistro     5.0
4                 4. Seaside Provisions     5.0
...                                 ...     ...
1220                 1046. Boston Pizza     3.5
1221            1047. Mirchi Restaurant     4.0
1222      1048. Beach Ave Bar and Grill     5.0
1223                   1049. Say Mercy!     4.5
1224                  1050. Tokyo Joe's     4.0

[1225 rows x 2 columns]

In [25]:
# delete ads reataurants (wo rate numbers)
df_rating = df_rating[df_rating['restaurant'].str.contains(r'\d')]
df_rating

restaurant  rating
1     1. Hydra Estiatorio Mediterranean     5.0
2                       2. Black + Blue     4.5
3                    3. Alouette Bistro     5.0
4                 4. Seaside Provisions     5.0
5           5. Salmon n' Bannock Bistro     4.5
...                                 ...     ...
1220                 1046. Boston Pizza     3.5
1221            1047. Mirchi Restaurant     4.0
1222      1048. Beach Ave Bar and Grill     5.0
1223                   1049. Say Mercy!     4.5
1224                  1050. Tokyo Joe's     4.0

[1050 rows x 2 columns]

Delete rating numbers for the all restaurants.

In [26]:
df_rating['restaurant'] = df_rating['restaurant'].str.replace(r'^\d+\. ', '',regex=True)

/var/folders/0z/fdtwx1zs7lg98s4zb6q_8_4h0000gn/T/ipykernel_4151/3924685758.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_rating['restaurant'] = df_rating['restaurant'].str.replace(r'^\d+\. ', '',regex=True)


### Merge

We will merge two clean dataframes based on the restaurant name.

In [30]:
df = df_review.merge(df_rating, on='restaurant', how='left' )
df

restaurant  \
0      Hydra Estiatorio Mediterranean   
1      Hydra Estiatorio Mediterranean   
2      Hydra Estiatorio Mediterranean   
3      Hydra Estiatorio Mediterranean   
4      Hydra Estiatorio Mediterranean   
...                               ...   
18442         Beach Ave Bar and Grill   
18443         Beach Ave Bar and Grill   
18444         Beach Ave Bar and Grill   
18445         Beach Ave Bar and Grill   
18446         Beach Ave Bar and Grill   

                                                  review  rating  
0      Wonderful fresh exotic Greek food that is deli...     5.0  
1      Food was delicious, packed full of flavors. Th...     5.0  
2      Hello, We would like to say thank you for the ...     5.0  
3      Happy hour was amazing. Buck a shuck oysters a...     5.0  
4      Roman was the best waiter! he checked up on us...     5.0  
...                                                  ...     ...  
18442  Stopped in for lunch while I was walking the S...     5.0  
18443  Great food and beer in a perfect spot looking ...     5.0  
18444  We dine here often. And I almost always get a ...     5.0  
18445  The Beach Ave bar and grill is lovely spot to ...     5.0  
18446  My wife & I, who live in northern BC, rented a...     5.0  

[18447 rows x 3 columns]

In [31]:
# check unique values
df.nunique()

restaurant     1009
review        16606
rating            4
dtype: int64

In [32]:
# check NaN values
df.isna().sum()

restaurant     0
review         0
rating        45
dtype: int64

Let's determine the number of restaurants that do not have a rating.

In [33]:
missing_rating_df = df.loc[df['rating'].isna()]
missing_rating_df['restaurant'].nunique

<bound method IndexOpsMixin.nunique of 15586    Ki-Isu Japanese Restaurant
15587    Ki-Isu Japanese Restaurant
15588    Ki-Isu Japanese Restaurant
15589    Ki-Isu Japanese Restaurant
15590    Ki-Isu Japanese Restaurant
15591    Ki-Isu Japanese Restaurant
15592    Ki-Isu Japanese Restaurant
15593    Ki-Isu Japanese Restaurant
15594    Ki-Isu Japanese Restaurant
15595    Ki-Isu Japanese Restaurant
15596    Ki-Isu Japanese Restaurant
15597    Ki-Isu Japanese Restaurant
15598    Ki-Isu Japanese Restaurant
15599    Ki-Isu Japanese Restaurant
15600    Ki-Isu Japanese Restaurant
15944     Won More Szechuan Cuisine
15945     Won More Szechuan Cuisine
15946     Won More Szechuan Cuisine
15947     Won More Szechuan Cuisine
15948     Won More Szechuan Cuisine
15949     Won More Szechuan Cuisine
15950     Won More Szechuan Cuisine
15951     Won More Szechuan Cuisine
15952     Won More Szechuan Cuisine
15953     Won More Szechuan Cuisine
15954     Won More Szechuan Cuisine
15955     Won More Szechu

Let's find the ratings of 3 restaurants and fill them manually. 

In [34]:
# fill missing rating for 45 restaurants

restaurant_rating_map = {
    'Ki-Isu Japanese Restaurant': 4.5,
    'Won More Szechuan Cuisine': 4,
    'Johnny’s On Oak': 3.5,}

df['rating'] = df['rating'].fillna(df['restaurant'].map(restaurant_rating_map))



In [35]:
# sanity check
df.isna().sum()

restaurant    0
review        0
rating        0
dtype: int64

There are not NaN values.

### Restaurant info

In [36]:
df_info = pd.read_csv('/Users/evgenijkucukov/Desktop/Brainstation/Portfolio/Restaurant_reviews_project/df_type')
df_info

restaurant           type
0       Freshslice Pizza        €2 - €5
1       Freshslice Pizza          Pizza
2       Freshslice Pizza  Lunch, Dinner
3     3. Alouette Bistro       €8 - €34
4     3. Alouette Bistro         French
...                  ...            ...
3124   1050. Tokyo Joe's  Lunch, Dinner
3125   1050. Tokyo Joe's        Takeout
3126    Freshslice Pizza        €2 - €5
3127    Freshslice Pizza          Pizza
3128    Freshslice Pizza  Lunch, Dinner

[3129 rows x 2 columns]

Let's repeat the same cleaning steps which we completed with reviews and rating.

In [37]:
# delete ads reataurants (wo rate numbers)
df_info = df_info[df_info['restaurant'].str.contains(r'\d')]
df_info

restaurant                                      type
3                3. Alouette Bistro                                  €8 - €34
4                3. Alouette Bistro                                    French
5                3. Alouette Bistro  Breakfast, Lunch, Dinner, Brunch, Drinks
6             4. Seaside Provisions                                  €8 - €41
7             4. Seaside Provisions                      Seafood, Dining bars
...                             ...                                       ...
3121  1049. Beach Ave Bar and Grill           Brew Pub, Pub, Beer restaurants
3122  1049. Beach Ave Bar and Grill             Lunch, Dinner, Brunch, Drinks
3123              1050. Tokyo Joe's                    Japanese, Sushi, Asian
3124              1050. Tokyo Joe's                             Lunch, Dinner
3125              1050. Tokyo Joe's                                   Takeout

[2577 rows x 2 columns]

Delete rating numbers for all restaurants.

In [38]:
df_info['restaurant'] = df_info['restaurant'].str.replace(r'^\d+\. ', '',regex=True)
df_info

/var/folders/0z/fdtwx1zs7lg98s4zb6q_8_4h0000gn/T/ipykernel_4151/4117296768.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_info['restaurant'] = df_info['restaurant'].str.replace(r'^\d+\. ', '',regex=True)


restaurant                                      type
3             Alouette Bistro                                  €8 - €34
4             Alouette Bistro                                    French
5             Alouette Bistro  Breakfast, Lunch, Dinner, Brunch, Drinks
6          Seaside Provisions                                  €8 - €41
7          Seaside Provisions                      Seafood, Dining bars
...                       ...                                       ...
3121  Beach Ave Bar and Grill           Brew Pub, Pub, Beer restaurants
3122  Beach Ave Bar and Grill             Lunch, Dinner, Brunch, Drinks
3123              Tokyo Joe's                    Japanese, Sushi, Asian
3124              Tokyo Joe's                             Lunch, Dinner
3125              Tokyo Joe's                                   Takeout

[2577 rows x 2 columns]

As observed, we have three rows of data for each restaurant, each containing different types of information. Let's consolidate all this information into a single row, separated by commas.

In [39]:
df_info_1 = df_info.groupby('restaurant')['type'].apply(','.join).reset_index()
df_info_1

Now, let's split each restaurant's characteristics, which are currently separated by commas, into separate columns.

In [40]:
# Find the maximum number of elements in any string
max_elements = df_info_1['type'].str.count(',') + 1
max_columns = max_elements.max()

# Split the 'type' column into a list of elements
df_info_1['elements'] = df_info_1['type'].str.split(',')

# Expand the list of elements into separate columns
for i in range(max_columns):
    df_info_1[f'col{i+1}'] = df_info_1['elements'].apply(lambda x: x[i] if i < len(x) else None)

# Drop the unnecessary columns
df_info_1.drop(columns=['type', 'elements'], inplace=True)

df_info_1

restaurant       col1  \
0    1927 Lobby Lounge at Rosewood Hotel Georgia  €26 - €34   
1                            1931 Gallery Bistro  €14 - €24   
2                    4 Stones Vegetarian Cuisine    Chinese   
3                             A La Mode Pie Café      Soups   
4                    A Taste Of India Restaurant     Indian   
..                                           ...        ...   
892         iDen & Quan Ju De Beijing Duck House    Chinese   
893                             sushi California   Japanese   
894                                  tetsu Sushi   Japanese   
895                                    the apron   €3 - €21   
896                                   the change   €7 - €34   

                    col2            col3                 col4  \
0                    Bar        Canadian             Wine Bar   
1               Canadian    Contemporary                 None   
2                  Asian       Taiwanese                 None   
3                   Cafe        Canadian            Breakfast   
4    Vegetarian Friendly   Vegan Options                Lunch   
..                   ...             ...                  ...   
892             European           Asian             Canadian   
893                Sushi           Asian  Vegetarian Friendly   
894                Sushi           Asian              Seafood   
895             Canadian   International  Vegetarian Friendly   
896              Chinese           Asian               Fusion   

                     col5               col6             col7  \
0                   Lunch             Dinner           Brunch   
1                    None               None             None   
2                    None               None             None   
3                   Lunch             Dinner           Brunch   
4                  Dinner         Late Night             None   
..                    ...                ...              ...   
892             Malaysian   Central European   South American   
893   Gluten Free Options              Lunch           Dinner   
894   Vegetarian Friendly             Dinner             None   
895   Gluten Free Options               None             None   
896                 Lunch             Dinner             None   

                  col8                    col9  ...           col29  \
0               Drinks                    None  ...            None   
1                 None                    None  ...            None   
2                 None                    None  ...            None   
3              Takeout   Wheelchair Accessible  ...            None   
4                 None                    None  ...            None   
..                 ...                     ...  ...             ...   
892   Central American         Beijing cuisine  ...   Table Service   
893               None                    None  ...            None   
894               None                    None  ...            None   
895               None                    None  ...            None   
896               None                    None  ...            None   

         col30                     col31         col32                  col33  \
0         None                      None          None                   None   
1         None                      None          None                   None   
2         None                      None          None                   None   
3         None                      None          None                   None   
4         None                      None          None                   None   
..         ...                       ...           ...                    ...   
892   Jazz Bar   Non-smoking restaurants   Sports bars   Gift Cards Available   
893       None                      None          None                   None   
894       None                      None          None                   None   
895       None                      None          None  

In [41]:
# check which columns have more values
df_info_1.nunique()

restaurant    897
col1          208
col2          104
col3           84
col4           78
col5           69
col6           57
col7           54
col8           47
col9           36
col10          34
col11          31
col12          25
col13          24
col14          19
col15          17
col16          16
col17          12
col18          12
col19          11
col20           7
col21           8
col22           8
col23           7
col24           5
col25           6
col26           4
col27           4
col28           4
col29           3
col30           3
col31           3
col32           2
col33           2
col34           1
col35           1
col36           1
col37           1
col38           1
dtype: int64

It's evident that columns 1 to 4 contain a substantial amount of relevant information about the restaurants. Therefore, let's proceed to remove the remaining columns.

In [42]:
# Drop columns
columns_to_drop = ['col38', 'col37', 'col36', 'col35', 'col34', 'col33', 'col32',
                   'col31', 'col30', 'col29', 'col28', 'col27', 'col26', 'col25',
                   'col24', 'col25', 'col24', 'col23', 'col22', 'col21', 'col20',
                   'col19', 'col18', 'col17', 'col16', 'col15', 'col14', 'col13', 
                   'col12', 'col11', 'col10', 'col9', 'col8', 'col7', 'col6',
                   'col5']
df_info_1.drop(columns=columns_to_drop, inplace=True)

df_info_1

restaurant       col1  \
0    1927 Lobby Lounge at Rosewood Hotel Georgia  €26 - €34   
1                            1931 Gallery Bistro  €14 - €24   
2                    4 Stones Vegetarian Cuisine    Chinese   
3                             A La Mode Pie Café      Soups   
4                    A Taste Of India Restaurant     Indian   
..                                           ...        ...   
892         iDen & Quan Ju De Beijing Duck House    Chinese   
893                             sushi California   Japanese   
894                                  tetsu Sushi   Japanese   
895                                    the apron   €3 - €21   
896                                   the change   €7 - €34   

                    col2            col3                 col4  
0                    Bar        Canadian             Wine Bar  
1               Canadian    Contemporary                 None  
2                  Asian       Taiwanese                 None  
3                   Cafe        Canadian            Breakfast  
4    Vegetarian Friendly   Vegan Options                Lunch  
..                   ...             ...                  ...  
892             European           Asian             Canadian  
893                Sushi           Asian  Vegetarian Friendly  
894                Sushi           Asian              Seafood  
895             Canadian   International  Vegetarian Friendly  
896              Chinese           Asian               Fusion  

[897 rows x 5 columns]

In [46]:
# combine restaurant' characteristics in one column 
df_info_1['Cuisine'] = df_info_1['col1'].str.cat([df_info_1['col2'], df_info_1['col3']], sep=', ')
df_info_1

restaurant       col1  \
0    1927 Lobby Lounge at Rosewood Hotel Georgia  €26 - €34   
1                            1931 Gallery Bistro  €14 - €24   
2                    4 Stones Vegetarian Cuisine    Chinese   
3                             A La Mode Pie Café      Soups   
4                    A Taste Of India Restaurant     Indian   
..                                           ...        ...   
892         iDen & Quan Ju De Beijing Duck House    Chinese   
893                             sushi California   Japanese   
894                                  tetsu Sushi   Japanese   
895                                    the apron   €3 - €21   
896                                   the change   €7 - €34   

                    col2            col3                 col4  \
0                    Bar        Canadian             Wine Bar   
1               Canadian    Contemporary                 None   
2                  Asian       Taiwanese                 None   
3                   Cafe        Canadian            Breakfast   
4    Vegetarian Friendly   Vegan Options                Lunch   
..                   ...             ...                  ...   
892             European           Asian             Canadian   
893                Sushi           Asian  Vegetarian Friendly   
894                Sushi           Asian              Seafood   
895             Canadian   International  Vegetarian Friendly   
896              Chinese           Asian               Fusion   

                                         Cuisine  
0                      €26 - €34, Bar,  Canadian  
1             €14 - €24, Canadian,  Contemporary  
2                    Chinese,  Asian,  Taiwanese  
3                        Soups,  Cafe,  Canadian  
4    Indian, Vegetarian Friendly,  Vegan Options  
..                                           ...  
892                   Chinese,  European,  Asian  
893                     Japanese,  Sushi,  Asian  
894                     Japanese,  Sushi,  Asian  
895           €3 - €21, Canadian,  International  
896                    €7 - €34, Chinese,  Asian  

[897 rows x 6 columns]

Let's delete col1, 2, 3, 4.

In [47]:
df_type = df_info_1.drop(columns = ['col1', 'col2', 'col3', 'col4'])

In [48]:
df_type

restaurant  \
0    1927 Lobby Lounge at Rosewood Hotel Georgia   
1                            1931 Gallery Bistro   
2                    4 Stones Vegetarian Cuisine   
3                             A La Mode Pie Café   
4                    A Taste Of India Restaurant   
..                                           ...   
892         iDen & Quan Ju De Beijing Duck House   
893                             sushi California   
894                                  tetsu Sushi   
895                                    the apron   
896                                   the change   

                                         Cuisine  
0                      €26 - €34, Bar,  Canadian  
1             €14 - €24, Canadian,  Contemporary  
2                    Chinese,  Asian,  Taiwanese  
3                        Soups,  Cafe,  Canadian  
4    Indian, Vegetarian Friendly,  Vegan Options  
..                                           ...  
892                   Chinese,  European,  Asian  
893                     Japanese,  Sushi,  Asian  
894                     Japanese,  Sushi,  Asian  
895           €3 - €21, Canadian,  International  
896                    €7 - €34, Chinese,  Asian  

[897 rows x 2 columns]

### Merge

Let's merge restaurant info with reviews and rating.

In [49]:
# merger 2 tables (review and rating with restaurant info) on restaurant name

df_final = df.merge(df_type, on='restaurant', how='left' )
df_final

restaurant  \
0      Hydra Estiatorio Mediterranean   
1      Hydra Estiatorio Mediterranean   
2      Hydra Estiatorio Mediterranean   
3      Hydra Estiatorio Mediterranean   
4      Hydra Estiatorio Mediterranean   
...                               ...   
18442         Beach Ave Bar and Grill   
18443         Beach Ave Bar and Grill   
18444         Beach Ave Bar and Grill   
18445         Beach Ave Bar and Grill   
18446         Beach Ave Bar and Grill   

                                                  review  rating  \
0      Wonderful fresh exotic Greek food that is deli...     5.0   
1      Food was delicious, packed full of flavors. Th...     5.0   
2      Hello, We would like to say thank you for the ...     5.0   
3      Happy hour was amazing. Buck a shuck oysters a...     5.0   
4      Roman was the best waiter! he checked up on us...     5.0   
...                                                  ...     ...   
18442  Stopped in for lunch while I was walking the S...     5.0   
18443  Great food and beer in a perfect spot looking ...     5.0   
18444  We dine here often. And I almost always get a ...     5.0   
18445  The Beach Ave bar and grill is lovely spot to ...     5.0   
18446  My wife & I, who live in northern BC, rented a...     5.0   

                         Cuisine  
0                            NaN  
1                            NaN  
2                            NaN  
3                            NaN  
4                            NaN  
...                          ...  
18442  €10 - €27, Brew Pub,  Pub  
18443  €10 - €27, Brew Pub,  Pub  
18444  €10 - €27, Brew Pub,  Pub  
18445  €10 - €27, Brew Pub,  Pub  
18446  €10 - €27, Brew Pub,  Pub  

[18447 rows x 4 columns]

In [50]:
df_final.nunique()

restaurant     1009
review        16606
rating            4
Cuisine         532
dtype: int64

In [51]:
df_final.to_csv('df_final_table', index=False)

For further analysis we have 1009 restaurants with rating, reviews and information about them.